plugins/sonarqube-backend/src/service/sonarqubeInfoProvider.ts

```diff
@@ -0,0 +1,371 @@
+    if (!(unnamedAllPresent || unnamedNonePresent)) {
+      throw new Error(
+        `Found partial default sonarqube config. All (or none) of baseUrl and apiKey must be provided.`,
+      );
+    }
+
+    if (unnamedAllPresent) {
+      const unnamedInstanceConfig = [
+        { name: DEFAULT_SONARQUBE_NAME, baseUrl, apiKey },
+      ] as {
+        name: string;
+        baseUrl: string;
+        apiKey: string;
+      }[];
+
+      return new SonarqubeConfig([
+        ...namedInstanceConfig,
+        ...unnamedInstanceConfig,
+      ]);
+    }
+
+    return new SonarqubeConfig(namedInstanceConfig);
+  }
+
+  /**
+   * Gets a Sonarqube instance configuration by name, or the default one if no name is provided.
+   * @param sonarqubeName - Optional name of the Sonarqube instance.
+   * @returns The requested Sonarqube instance.
+   * @throws Error when no default config could be found or the requested name couldn't be found in config.
+   */
+  getInstanceConfig(sonarqubeName?: string): SonarqubeInstanceConfig {
+    const DEFAULT_SONARQUBE_NAME = 'default';
+
+    if (!sonarqubeName || sonarqubeName === DEFAULT_SONARQUBE_NAME) {
+      // no name provided, use default
+      const instanceConfig = this.instances.find(
+        c => c.name === DEFAULT_SONARQUBE_NAME,
+      );
+
+      if (!instanceConfig) {
+        throw new Error(
+          `Couldn't find a default sonarqube instance in the config. Either configure an instance with name ${DEFAULT_SONARQUBE_NAME} or add a prefix to your annotation value.`,
+        );
+      }
+
+      return instanceConfig;
+    }
+
+    // A name is provided, look it up.
+    const instanceConfig = this.instances.find(c => c.name === sonarqubeName);
+
+    if (!instanceConfig) {
+      throw new Error(
+        `Couldn't find a sonarqube instance in the config with name ${sonarqubeName}`,
+      );
+    }
+    return instanceConfig;
+  }
+}
+
+/**
+ * @public
+ *
+ * Use default config and annotations, build using fromConfig static function.
+ */
+export class DefaultSonarqubeInfoProvider implements SonarqubeInfoProvider {
+  private constructor(private readonly config: SonarqubeConfig) {}
+
+  /**
+   * Generate an instance from a Config instance
+   * @param config - Backend configuration
+   */
+  static fromConfig(config: Config): DefaultSonarqubeInfoProvider {
+    return new DefaultSonarqubeInfoProvider(SonarqubeConfig.fromConfig(config));
+  }
+
+  /**
+   * Retrieve all supported metrics from a sonarqube instance.
+   *
+   * @param instanceUrl - URL of the sonarqube instance
+   * @param token - token to access the sonarqube instance
+   * @returns The list of supported metrics, if no metrics are supported an empty list is provided in the promise
+   * @private
+   */
+  private static async getSupportedMetrics(
+    instanceUrl: string,
+    token: string,
+  ): Promise<string[]> {
+    const metrics: string[] = [];
+    let nextPage: number = 1;
+
+    for (;;) {
```

Comment:
I wonder if it make sense to have a hard limit here too?

Reply:
Nah, I think it's fine, if it's just a lift and shift it's fine we can revisit later.